register/register.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{
+		"access_token":  t,
+		"refresh_token": rt,
+	}, nil
+}
+
+func New(db *sqlx.DB) *RegisterRepository {
+	return &RegisterRepository{Db: db}
+}
+
+func (repository *RegisterRepository) Register(c *gin.Context) {
+
+	input := UserRegister{}
+	err := c.ShouldBindWith(&input, binding.JSON)
+	var email string
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"Message ": err.Error()})
+		return
+	}
+
+	err = repository.Db.Get(&email, `SELECT email from Users where email=?`, input.Email)
+
+	if email != input.Email {
+
+		auth := generateRefreshToken()
+
+		_, err = repository.Db.Exec(`INSERT INTO Users(name,email,user_type,auth_token) VALUES (?,?,?,?)`, input.Name, input.Email, "General", auth)
+
+		if err != nil {
+			panic(err)
+		}
+
+		c.JSON(http.StatusOK, gin.H{"message": "Register Successfully"})
+		c.JSON(http.StatusOK, auth)
+
+	} else {
+		c.JSON(403, gin.H{"message": "Email Already exist"})
+	}
+
+}
+
+// ADMIN REGISTER
+func (repository *RegisterRepository) RegisterAdmin(c *gin.Context) {
+
+	input := UserRegister{}
+	var email string
+
+	err := c.ShouldBindWith(&input, binding.JSON)
+
+	if err != nil {
+		c.Abort()
+		c.JSON(http.StatusBadRequest, gin.H{"Message ": err.Error()})
+		return
+	}
+
+	err = repository.Db.Get(&email, `SELECT email from Users where email= ?`, input.Email)
+
+	if email != input.Email {
+
+		auth := generateRefreshToken()
+
+		_, err = repository.Db.Exec(`INSERT INTO Users(name,email,user_type,auth_token) VALUES (?,?,?,?)`, input.Name, input.Email, "Admin", auth)
+
+		c.JSON(http.StatusOK, gin.H{"message": "Admin Registered Successfully"})
+		c.JSON(http.StatusOK, auth)
+
+	} else {
+		c.JSON(403, gin.H{"message": "Email Already exist"})
+	}
+}
+
+//User Login
+
+func (repository *RegisterRepository) Login(c *gin.Context) {
+
+	type tokenReqBody struct {
+		RefreshToken string `json:"refresh_token"`
+		AccessToken  string `json:"access_token"`
+	}
+
+	tokenReq := tokenReqBody{}
+	c.Bind(&tokenReq)
+
+	input := UserRegister{}
+	token := Token{}
+	var email string
+
+	if err := c.ShouldBindJSON(&input); err != nil {
+		c.JSON(http.StatusUnprocessableEntity, "Invalid json provided")
+
+	}
+
+	err := repository.Db.Get(&email, `SELECT email FROM Users WHERE email= ?`, input.Email)
+	if err != nil {
+		panic(err)
+	}
+	//compare the user from the request, with the one we defined:
+	if email != input.Email {
+		c.JSON(http.StatusUnauthorized, "Please provide valid login details")
+		return
+	}
+	err = repository.Db.Get(&token, `SELECT email,user_type FROM Users WHERE email= ?`, input.Email)
+	if err != nil {
+		panic(err)
+	}
+
+	validToken, err := GenerateJWT(token.Email, token.Role)
+
+	c.JSON(http.StatusOK, validToken)
+
+	if err != nil {
+		panic(err)
+		return
+	}
+
+	var tokenDetail Token
+	tokenDetail.Email = token.Email
+	tokenDetail.Role = token.Role
+	tokenDetail.TokenString = validToken["access_token"]
```

Comment:
Why do we need to set token detail?

Reply:
i have used to check whether i was getting proper data or not ..i will remove it .